src/interface/sparse_pack_base.cpp

```diff
@@ -0,0 +1,265 @@
+                  idx++;
+                }
+              }
+            }
+          }
+        }
+      }
+
+      bounds_h(1, b, i) = idx - 1;
+
+      if (bounds_h(1, b, i) < bounds_h(0, b, i)) {
+        // Did not find any allocated variables meeting our criteria
+        bounds_h(0, b, i) = -1;
+        // Make the upper bound more negative so a for loop won't iterate once
+        bounds_h(1, b, i) = -2;
+      }
+    }
+  });
+
+  Kokkos::deep_copy(pack.pack_, pack_h);
+  Kokkos::deep_copy(pack.bounds_, bounds_h);
+  Kokkos::deep_copy(pack.coords_, coords_h);
+  pack.ndim_ = ndim;
+  pack.dims_[1] = pack.nblocks_;
+  pack.dims_[2] = -1; // Not allowed to ask for the ragged dimension anyway
+  pack.dims_[3] = pack_h(0, 0, 0).extent_int(0);
+  pack.dims_[4] = pack_h(0, 0, 0).extent_int(2);
+  pack.dims_[5] = pack_h(0, 0, 0).extent_int(3);
+
+  return pack;
+}
+
+// Specialize for the only two types this should work for
+template SparsePackBase
+SparsePackBase::Build<MeshBlockData<Real>>(MeshBlockData<Real> *, const PackDescriptor &);
+template SparsePackBase SparsePackBase::Build<MeshData<Real>>(MeshData<Real> *,
+                                                              const PackDescriptor &);
+
+inline SparsePackBase::test_func_t
+SparsePackBase::GetTestFunction(const PackDescriptor &desc) {
+  std::vector<std::regex> regexes;
+  for (const auto &var : desc.vars)
+    regexes.push_back(std::regex(var));
+
+  // Lambda for testing wether or not we want to include cellvariable pv
+  // in the index range of the type variable with index vidx
+  return [=](int vidx, const std::shared_ptr<CellVariable<Real>> &pv) {
+    // TODO(LFR): Check that the shapes agree
+
+    if (desc.flags.size() > 0) {
+      for (const auto &flag : desc.flags) {
+        if (!pv->IsSet(flag)) {
+          return false;
+        }
+      }
+    }
+
+    if (desc.use_regex[vidx]) {
+      if (std::regex_match(std::string(pv->label()), regexes[vidx])) return true;
+    } else {
+      if (desc.vars[vidx] == pv->label()) return true;
+    }
+    return false;
+  };
+}
+
+template <class T>
+SparsePackBase &SparsePackCache::Get(T *pmd, const PackDescriptor &desc) {
+  std::string ident = GetIdentifier(desc);
+  if (pack_map.count(ident) > 0) {
+    auto &pack = pack_map[ident].first;
+    if (desc.with_fluxes != pack.with_fluxes_) goto make_new_pack;
+    if (desc.coarse != pack.coarse_) goto make_new_pack;
+    auto alloc_status_in = SparsePackBase::GetAllocStatus(pmd, desc);
+    auto &alloc_status = pack_map[ident].second;
+    if (alloc_status.size() != alloc_status_in.size()) goto make_new_pack;
+    for (int i = 0; i < alloc_status_in.size(); ++i) {
+      if (alloc_status[i] != alloc_status_in[i]) goto make_new_pack;
+    }
+    return pack_map[ident].first;
+  }
+
+make_new_pack:
+  pack_map[ident] = {SparsePackBase::Build(pmd, desc),
+                     SparsePackBase::GetAllocStatus(pmd, desc)};
+  return pack_map[ident].first;
+}
```

Comment:
Is there a way we could avoid the `goto`s here?

Reply:
https://imgflip.com/i/6mtqtc
To be fair it would be pretty easy to remove the `goto`s by defining a new function 
```c++
template <class T>
SparsePackBase &MakeNewPack(T *pmd, const PackDescriptor &desc) { 
  pack_map[ident] = {SparsePackBase::Build(pmd, desc),
                     SparsePackBase::GetAllocStatus(pmd, desc)};
  return pack_map[ident].first;
}
```
and replacing all of the `goto`s with `return MakeNewPack(pad, desc);`